File: frontend/src/lib/content.ts
import { useQuery } from "@tanstack/react-query";
import { apiGet } from "@/lib/api";
import { IMG, SKILL_CLUSTERS, TOOLS } from "@/lib/data";

export interface SkillCluster {
  title: string;
  items: string[];
}

export interface ExperienceImages {
  shalom: string;
  moustache: string;
}

export interface ContactInfo {
  headingA: string;
  headingB: string;
  intro: string;
  email: string;
  linkedin: string;
  whatsapp: string;
  basedIn: string;
  focus: string;
  footerTagline: string;
}

export interface SiteContent {
  skills: SkillCluster[];
  toolkit: string[];
  categories: string[];
  experience: ExperienceImages;
  contact: ContactInfo;
}

export const DEFAULT_CATEGORIES = [
  "Photoshoot",
  "Website & CRM",
  "Graphic Design / Content",
  "Events",
  "Influencer Collab",
  "Listings",
];

export const DEFAULT_CONTACT_INFO: ContactInfo = {
  headingA: "Have a travel, hospitality",
  headingB: "or marketing project?",
  intro:
    "I'm open to opportunities in tourism, hospitality marketing, travel content, freelance projects and travel product building.",
  email: "hello@vaibhavkanhere.com",
  linkedin: "https://www.linkedin.com/in/vaibhav-kanhere",
  whatsapp: "https://wa.me/910000000000",
  basedIn: "India · open to remote & on-site",
  focus: "Tourism & hospitality marketing · content · travel systems",
  footerTagline: "I market travel businesses — and I build the systems behind them.",
};

const DEFAULT_CONTENT: SiteContent = {
  skills: SKILL_CLUSTERS.map((c) => ({ title: c.title, items: [...c.items] })),
  toolkit: [...TOOLS],
  categories: [...DEFAULT_CATEGORIES],
  experience: { shalom: IMG.rishikeshRiver, moustache: IMG.chefPlating },
  contact: DEFAULT_CONTACT_INFO,
};

export function useSiteContent() {
  const q = useQuery({
    queryKey: ["site-content"],
    queryFn: () => apiGet<Partial<SiteContent>>("/content"),
    staleTime: 30_000,
    retry: 1,
  });
  const content: SiteContent = {
    skills: q.data?.skills?.length ? (q.data.skills as SkillCluster[]) : DEFAULT_CONTENT.skills,
    toolkit: q.data?.toolkit?.length ? q.data.toolkit : DEFAULT_CONTENT.toolkit,
    categories: q.data?.categories?.length ? q.data.categories : DEFAULT_CONTENT.categories,
    experience: q.data?.experience?.shalom
      ? (q.data.experience as ExperienceImages)
      : DEFAULT_CONTENT.experience,
    contact: q.data?.contact?.email
      ? { ...DEFAULT_CONTACT_INFO, ...(q.data.contact as Partial<ContactInfo>) }
      : DEFAULT_CONTENT.contact,
  };
  return { content, isError: q.isError };
}

File: backend/routers/content.py
from typing import List

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from lib.admin_auth import require_admin
from lib.db import db

router = APIRouter()

DEFAULT_CATEGORIES = [
    "Photoshoot",
    "Website & CRM",
    "Graphic Design / Content",
    "Events",
    "Influencer Collab",
    "Listings",
]


class SkillCluster(BaseModel):
    title: str
    items: List[str]


class SiteContent(BaseModel):
    skills: List[SkillCluster] = []
    toolkit: List[str] = []
    categories: List[str] = []
    experience: dict = {}


async def get_allowed_categories() -> set[str]:
    """Union of the built-in defaults and any categories the admin added via the dashboard."""
    doc = await db.site_content.find_one({"key": "main"}, {"_id": 0})
    stored = ((doc or {}).get("value") or {}).get("categories") or []
    return set(DEFAULT_CATEGORIES) | {c.strip() for c in stored if c.strip()}


@router.get("/content")
async def get_content():
    doc = await db.site_content.find_one({"key": "main"}, {"_id": 0})
    return doc["value"] if doc else {}


@router.put("/content", dependencies=[Depends(require_admin)])
async def put_content(body: SiteContent):
    await db.site_content.update_one({"key": "main"}, {"$set": {"value": body.model_dump()}}, upsert=True)
    return {"ok": True}
